### ISSUE TITLE
sql: deadlock with auto stats on a cluster shutdown

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#99753]
On [this](https://teamcity.cockroachdb.com/viewLog.html?buildId=9281304&buildTypeId=Cockroach_BazelEssentialCi) (also seen [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=9281541&buildTypeId=Cockroach_BazelEssentialCi)) CI run on staging branch

<details>
<summary>goroutines</summary>

```
goroutine 240420 [select, 59 minutes]:
github.com/cockroachdb/cockroach/pkg/sql.(*ieResultChannel).firstResult(0xc0056dad80, {0x64df4e0, 0xc00582fd70})
	github.com/cockroachdb/cockroach/pkg/sql/internal_result_channel.go:123 +0xf9
github.com/cockroachdb/cockroach/pkg/sql.(*ieResultChannel).nextResult(0xc0056dad80?, {0x64df4e0, 0xc00582fd70})
	github.com/cockroachdb/cockroach/pkg/sql/internal_result_channel.go:162 +0xfe
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal(0xc007bb26e0, {0x64df4e0, 0xc00582f710}, {0x50e24da, 0x8}, 0xc0056dad80, 0x0?, {{{0x0, 0x0}}, {0x0, ...}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1004 +0x1425
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryIteratorEx(0x203001?, {0x64df4e0, 0xc00582f710}, {0x50e24da, 0x8}, 0x0?, {{{0x0, 0x0}}, {0x0, 0x0}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:704 +0x1cb
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryIterator(0x0?, {0x64df4e0, 0xc00582f710}, {0x50e24da, 0x8}, 0x1?, {0x52c555e, 0xae}, {0x0, 0x0, ...})
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:690 +0x165
github.com/cockroachdb/cockroach/pkg/jobs.RunningJobExists({0x64df4e0, 0xc00582f710}, 0x0, {0x6531260, 0xc007bb26e0}, 0x54066a8)
	github.com/cockroachdb/cockroach/pkg/jobs/utils.go:42 +0xd5
github.com/cockroachdb/cockroach/pkg/sql.checkRunningJobs.func1({0x64df4e0?, 0xc00582f710?}, {0x6531260?, 0xc007bb26e0?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:711 +0x4a
github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1({0x64df4e0?, 0xc00582f710?}, 0x65407c0?)
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1422 +0x31
github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4({0x64df4e0, 0xc00582f710}, 0xc006ccefd0)
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1500 +0x3b4
github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1({0x64df4e0?, 0xc00582f710?}, 0xc00972e3c0?)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:965 +0x27
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc006ccefd0, {0x64df4e0, 0xc00582f710}, 0xc001173bd8)
	github.com/cockroachdb/cockroach/pkg/kv/txn.go:928 +0xae
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x64df4e0, 0xc00582f710}, 0xc00582f710?, 0x5b07650?)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:964 +0x6b
github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0x7f433fc72108?, {0x64df4e0, 0xc00582f710}, 0x56dad00?, 0xc0?, 0x0?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:927 +0xa7
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0x439247?, {0x64df4e0?, 0xc00582f710?}, 0x1?)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:902 +0x2d
github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn(0xc004756630, {0x64df4e0, 0xc00582f710}, 0xc002a58f30, {0x0, 0x0, 0x4bda0c0?})
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1488 +0x2a7
github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn(0xc00618f100?, {0x64df4e0, 0xc00582f710}, 0xc0058c8b28, {0x0, 0x0, 0x0})
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1423 +0x9f
github.com/cockroachdb/cockroach/pkg/sql.checkRunningJobs({0x64df4e0, 0xc00582f710}, 0xc00582f710?, {0x652c8b0, 0xc0086ab608})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:710 +0xd1
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startJob(0xc00383b500, {0x64df4e0, 0xc00582f710}, 0xc00582eb10?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:153 +0x96
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec.func1()
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:116 +0x4a
created by github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:115 +0x171

goroutine 240419 [select, 59 minutes]:
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).Next(0xc00383b500, {{0x64df4e0?, 0xc00582f710?}, 0xc004772800?, 0xc0086ab608?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:128 +0xc7
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc001e298c0)
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:217 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).next(0xc005c8bc20)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:129 +0x3b
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).nextAdapter(0xc005c8bc20)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:146 +0x1e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x2?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x62
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Next(0xc005c8bc20)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:151 +0x3e
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run({0x64df4e0, 0xc00582f6b0}, {0x6500de0, 0xc005c8bc20}, {0x64b6d40?, 0xc00528e700?})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:191 +0x55
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc005c8bc20, {0x64df438?, 0xc0056daac0?}, {0x64b6d40?, 0xc00528e700})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:737 +0x73
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc00b276280, {0x64df438?, 0xc0056daac0}, 0x20?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:552 +0x21a
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x6541de0?, {0x64df438?, 0xc0056daac0?}, 0x70?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:312 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc006dce280, {0x64df4e0, 0xc00582f4d0}, 0xc006ccef20, 0xc006ccee70, 0xc00383b580, 0xc00528e700, 0xc0086aba48, 0xc002a58d80)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:904 +0xb73
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x346318d?, {0x64df4e0, 0xc00582f4d0}, 0xc00383b500?, 0xc006ccef20, 0x203001?, {{0x64e0c80?, 0xc00383b500?}, 0x0?}, 0xc00528e700, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1900 +0x205
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func2(0xc0086ab608, 0x0?, {0x64df4e0, 0xc00582f4d0}, 0x0?, 0x0?, 0x10?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1636 +0xcb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0x0?, {0x64df4e0, 0xc00582f4d0}, 0xc0086aba48, 0x64df4e0?, 0xc0086ab608, 0xc00528e700, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1639 +0x1e5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc0086ab000, {0x64df4e0?, 0xc00582f4d0}, 0xc0086ab608, 0xc00582f4d0?, {0x6531360?, 0xc00582f0e0}, 0xc00582f0e0?, 0xc00356ce18)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1771 +0x5eb
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0086ab000, {0x64df438, 0xc0056daa00}, 0xc0086ab608, {0x6531360, 0xc00582f0e0})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1351 +0xc19
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0086ab000, {0x64df4e0?, 0xc00582f020?}, {{0x6507c20, 0xc0058f26c0}, {0x0, 0x0, 0x0}, {0xc0085f0120, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:770 +0x2c0f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x64df4e0?, 0xc00582f020?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:137 +0xbc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x64df4e0?, {0x64df4e0?, 0xc00582f020?}, {0x6507c20?, 0xc0058f26c0?}, 0x64df4e0?, 0xc00582f020?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2660 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0086ab000, {0x64df4e0, 0xc00582f020}, {{0x6507c20, 0xc0058f26c0}, {0x0, 0x0, 0x0}, {0xc0085f0120, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:136 +0x4da
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc004bb9bd0, 0xc0086ab000, 0xc004bb9890, 0xc004bb9950, 0xc004bb9990, 0xc004987970, 0xc004987960, 0xc004987980)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2118 +0x319
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc0086ab000)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2123 +0x1127
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0086ab000, {0x64df4e0, 0xc00582eb10}, 0x11ba840?, 0xc005deb400?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2039 +0x20b
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:172 +0xac
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:171 +0x195

goroutine 239178 [select, 59 minutes]:
github.com/cockroachdb/cockroach/pkg/sql.(*ieResultChannel).firstResult(0xc0056da7c0, {0x64df4e0, 0xc00582eb10})
	github.com/cockroachdb/cockroach/pkg/sql/internal_result_channel.go:123 +0xf9
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal(0xc009b1c640, {0x64df4e0, 0xc0089a91a0}, {0x50f00e7, 0xc}, 0xc0056da7c0, 0x528334?, {{{0x50d7c07, 0x4}}, {0x0, ...}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:988 +0x11c9
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).ExecEx(0x8eb4940?, {0x64df4e0, 0xc0089a91a0}, {0x50f00e7, 0xc}, 0xc004bb9c00?, {{{0x50d7c07, 0x4}}, {0x0, 0x0}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:666 +0x19b
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).Exec(0x4?, {0x64df4e0, 0xc0089a91a0}, {0x50f00e7, 0xc}, 0xedbb4134f?, {0xc0085f0120, 0x5a}, {0x0, 0x0, ...})
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:647 +0x165
github.com/cockroachdb/cockroach/pkg/sql/stats.(*Refresher).refreshStats(0xc005e5f0e0, {0x64df4e0, 0xc0089a91a0}, 0x2e2a3810?, 0xedbb4187e?)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats.go:765 +0x196
github.com/cockroachdb/cockroach/pkg/sql/stats.(*Refresher).maybeRefreshStats(0xc005e5f0e0, {0x64df4e0, 0xc0089a91a0}, 0xc, 0x71f7e8ff2?, 0x1, 0xcc0143c767776067?)
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats.go:733 +0x2f9
github.com/cockroachdb/cockroach/pkg/sql/stats.(*Refresher).Start.func1.1({0x64df4e0, 0xc0089a91a0})
	github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/automatic_stats.go:473 +0x4c8
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```

</details>

It seems that the internal `get-jobs` query was improperly canceled - one of the flows was canceled but nothing was communicated back to the `ieResultChannel`:
```
W230327 22:44:32.499991 240710 sql/colflow/colrpc/outbox.go:183 â‹® [T1,n1,intExec=â€¹system-jobs-scanâ€º,fâ€¹2820ac2bâ€º,distsql.stmt=â€¹WITH latestpayload AS (SELECT job_id, value FROM system.job_info AS payload WHERE info_key = '_'::BYTES), latestprogress AS (SELECT job_id, value FROM system.job_info AS progress WHERE info_key = '_'::BYTES) SELECT id, status, created, payload.value AS payload, progress.value AS progress, created_by_type, created_by_id, claim_session_id, claim_instance_id, num_runs, last_run, job_type FROM system.jobs AS j INNER JOIN latestpayload AS payload ON j.id = payload.job_id LEFT JOIN latestprogress AS pâ€º,distsql.gateway=â€¹1â€º,distsql.appname=â€¹$ internal-get-jobsâ€º,distsql.txn=â€¹bc46de0d-5f67-486f-9b1a-84e64f76d31fâ€º,streamID=â€¹20â€º] 936  Outbox Dial connection error, distributed query will fail: dial: context canceled
```
Perhaps a71e11c3137a3242a6780c70ab6c2062a31f504a is to blame.

Jira issue: CRDB-26053

**Comment by yuzefovich** [comment]
I've bisected this to 5d9b97a67c6bade8bb5a921b7fb881ac44f11579, cc @adityamaru. Unsure whether it's a pre-existing problem that was simply exposed or a new problem introduced in that commit.

**Comment by yuzefovich** [comment]
I think I tracked it down, and it's a pre-existing bug.

**Comment by yuzefovich** [comment]
Re-opening to track merging the fix to 23.1

**Comment by yuzefovich** [comment]
This was fixed in #99969 on 23.1.
